lib/Adapter/Tolerant/Indexer/ClassLikeReferenceIndexer.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace Phpactor\Indexer\Adapter\Tolerant\Indexer;
+
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\QualifiedName;
+use Phpactor\Indexer\Model\Index;
+use Phpactor\Indexer\Model\RecordReference;
+use Phpactor\Indexer\Model\Record\ClassRecord;
+use Phpactor\Indexer\Model\Record\FileRecord;
+use SplFileInfo;
+
+class ClassLikeReferenceIndexer extends AbstractClassLikeIndexer
+{
+    public function canIndex(Node $node): bool
+    {
+        return $node instanceof QualifiedName;
+    }
+
+    public function beforeParse(Index $index, SplFileInfo $info): void
+    {
+        $fileRecord = $index->get(FileRecord::fromFileInfo($info));
+        assert($fileRecord instanceof FileRecord);
+
+        foreach ($fileRecord->references() as $outgoingReference) {
+            if ($outgoingReference->type() !== ClassRecord::RECORD_TYPE) {
+                continue;
+            }
+
+            $record = $index->get(ClassRecord::fromName($outgoingReference->identifier()));
+            assert($record instanceof ClassRecord);
+            $record->removeReference($fileRecord->identifier());
+            $index->write($record);
+        }
+    }
```

Comment:
When parsing a file we can safely remove all references to the file, as they will be recreated as required.